File: tp-fizz-buzz-java/src/main/java/fr/isima/service/FizzBuzz.java
package fr.isima.service;

import java.util.SortedMap;
import java.util.TreeMap;

public class FizzBuzz {


    public String fizzBuzz(int numberToFizz) {
        final StringBuilder fizzBuzzBuilder = new StringBuilder();
        if (numberToFizz % 3 == 0) {
            fizzBuzzBuilder.append("Fizz");
        }
        if (numberToFizz % 5 == 0) {
            fizzBuzzBuilder.append("Buzz");
        }
        if (fizzBuzzBuilder.length() == 0) {
            fizzBuzzBuilder.append(numberToFizz);
        }
        return fizzBuzzBuilder.toString();
    }

    public FizzBuzzSet fizzBuzzAll(int start, int end) {
        final SortedMap<Integer, String> sorted = new TreeMap<>();
        for (int i = start; i <= end; i++) {
            sorted.put(i, fizzBuzz(i));
        }
        return new FizzBuzzSet(sorted);
    }
}

File: tp-fizz-buzz-kotlin/src/main/kotlin/fr/isima/service/FizzBuzz.kt
package fr.isima.service

/**
 * Type alias, in current implementation we just want to use an immutable map (map is immutable in kotlin)
 * In future, if we want to create a complex, can we will be able to do it
 */

typealias FizzBuzzSet = Map<Int, String>
/**
 * This is the main extension function solving
 */
val Int.fizzBuzz: String
    get() {
        //just an example of 'by lazy usage', here we could simply
        //initialize isFizz or isBuzz directly
        //but the examples shows how lambda can be smart
        val isFizz by lazy { this % 3 == 0 }
        val isBuzz by lazy { this % 5 == 0 }
        return when {
            isFizz && isBuzz -> "FizzBuzz"
            isFizz -> "Fizz"
            isBuzz -> "Buzz"
            else -> this.toString()
        }
    }

/**
 * Using property exension to add fizzBuzz to IntRange
 */
val IntRange.fizzBuzz: FizzBuzzSet
    get() = asIterable().fizzBuzz

/**
 * Iterable is a Java interface more general than Collection. It just means i can browse some values. A lot of Kotlin extensions functions are avaialable through Iterable.
 *
 *
 */
val Iterable<Int>.fizzBuzz: FizzBuzzSet
    get() = associate { it to it.fizzBuzz }.toSortedMap()


val FizzBuzzSet.plainText: String
    get() = when {
        isEmpty() -> "No fizz buzz !"
        else -> "FizzBuzz all !!!! :\n" + map { (key, value) -> "$key : $value" }.reduce { prev, next -> "$prev\n$next" }
    }

